jujugui/static/gui/src/app/app.js

```diff
@@ -856,6 +867,39 @@ YUI.add('juju-gui', function(Y) {
     },
 
     /**
+      Handles redirecting to a new model when attempting to log in.
+
+      @method _handleRedirectionRequired
+      @param {String} err The error message string.
+      @param {Object} resp The response value containing the available server
+        details.
+    */
+    _handleRedirectionRequired: function(err, resp) {
+      // If we have a response and we have a redirection required error
+      // then we need to modify the connection information and try again.
+      // Otherwise we just carry on not doing anything.
+      if (resp && (err && err.indexOf('redirection required') > -1)) {
+        // Loop through the available servers and find the public IP.
+        const server = resp.servers[0].filter(
+          server => server.scope === 'public');
+        // The environmentList will be populated by the model listing
+        // request after the controllerAPI call has connected.
+        const selectedModel = this._pickModel(
+          this.get('environmentList'));
+        let socketTemplate = this.get('socketTemplate');
+        // When connecting to a model using JIMM and a redirection, the socket
+        // template base differs from the default so we must provide
+        // it here when creating the socket url.
+        if (this.get('jimmURL')) {
+          socketTemplate =
+            `wss://${server[0].value}:${server[0].port}${socketTemplate}`;
```

Comment:
So here the socket template should be set to wss://$host:$port... already, and then the real values rendered there because they are part of the context. Otherwise this will never work on wss proxy cases and it basically breaks the whole concept of socket templates, and the code relying to them.